Sidorina_Test_1

ElenaLaufeyson opened the pull request:
_Решила первый урок._

ElenaLaufeyson committed:
Весь решенный 1 тест.

ElenaLaufeyson committed:
Весь решенный 1 тест.

mglukhikh commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +96,33 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  * Вывести значение того же угла в радианах (например, 0.63256).
  */
-fun angleInRadian(grad: Int, min: Int, sec: Int): Double = TODO()
+fun angleInRadian(grad: Int, min: Int, sec: Int): Double = (Math.PI / 180)*(grad + min/60 + sec/3600)
```
min/60 и sec/3600 -- это целочисленное деление (нацело). А поскольку в примерах минут и секунд не больше 59, результат этого деления будет всегда равен 0. Чтобы операция деления имела вещественный результат, один из её аргументом должен быть вещественным. Проще всего этого добиться, добавив дробную часть константе.

mglukhikh commented on src/lesson1/task1/Simple.kt:
```diff
@@ -93,7 +131,8 @@ fun thirdDigit(number: Int): Int = TODO()
  */
-fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
+fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int):
+        Int = (minutesArrive + 60*hoursArrive) - (minutesDepart + 60*hoursDepart)
```
Обычно здесь перенос делается так:

```
fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = 
        (minutesArrive + 60*hoursArrive) - (minutesDepart + 60*hoursDepart)
```

Тогда заголовок функции у вас весь в одной строчке, а тело в другой -- лучше читается.

mglukhikh commented on test/lesson1/task1/Tests.kt:
```diff
@@ -54,7 +54,7 @@ class Tests {
     fun angleInRadian() {
-        assertEquals(0.63256, angleInRadian(36, 14, 35), 1e-5)
+        assertEquals(0.63256, angleInRadian(36, 14, 35), 1e-2)
         assertEquals(Math.PI / 2.0, angleInRadian(90, 0, 0), 1e-5)
```
Здесь неспроста была указана допустимая погрешность именно 1e-5. Вы её увеличили, и как результат не видите собственную ошибку (см. выше).

ElenaLaufeyson committed:
Исправила ошибки в первом уроке (синтаксис и неправильно решенную функцию angleInRadian), а также сделала 3 задания из второго урока.

ElenaLaufeyson committed:
Исправила ошибки из второго урока (сделаны только первые 3 задания).

ElenaLaufeyson committed:
Исправляю ошибки из второго урока...

mglukhikh commented:
Хорошо, первый урок засчитан.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +50,23 @@ fun ageDescription(age: Int): String = TODO()
+    val s1 = t1*v1
+    val s2 = t2*v2
+    val s3 = t3*v3
+    val HalfS = (s1 + s2 + s3)/2
```
Переменные не следует называть с большой буквы

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +78,17 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    return when {
```
Это решение можно записать существенно короче

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +50,23 @@ fun ageDescription(age: Int): String = TODO()
+    if (HalfS == 0.0) return Double.NaN
+    return when {
+        HalfS == s1 -> t1
+        HalfS == s2 -> t2
```
Это условие неверное

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +50,23 @@ fun ageDescription(age: Int): String = TODO()
+    val s2 = t2*v2
+    val s3 = t3*v3
+    val HalfS = (s1 + s2 + s3)/2
+    if (HalfS == 0.0) return Double.NaN
```
В задании явно не написано, каким должен быть результат, если весь пройденный путь равен нулю. Строго говоря, в подобной ситуации должен подходить любой ответ в интервале от 0 до t1 + t2 + t3, Double.NaN как признак неопределённости ответа тоже подходит. Поэтому, заявление бота о том, что ваш ответ в этой ситуации неверен -- неправильно.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +50,23 @@ fun ageDescription(age: Int): String = TODO()
+    return when {
+        HalfS == s1 -> t1
+        HalfS == s2 -> t2
+        (HalfS < s1) && (v1 != 0.0) -> HalfS/v1
```
А бывает так, что Halfs < s1 && v1 == 0.0? Если да, придумайте тест на эту ситуацию и добавьте его в тестовую функцию.

mglukhikh commented:
По второму уроку незначительные замечания по коду. Вы можете также, если чувствуете в себе силы, порешать более сложные задачи из второго урока

ElenaLaufeyson committed:
Lesson2_task1
Исправлены ошибки в функциях timeForHalfWay и whichRookThreatens и решена функция rookOrBishopThreatens

mglukhikh commented:
Ваши ответы в задаче про ладью и слона верные, сообщение бота ошибочно.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -101,7 +114,43 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+        {
+            if (rookX in min(bishopX, kingX)..max(bishopX, kingX) &&
+                    rookY in min(bishopY, kingY)..max(bishopY, kingY))
+                dangerbishop = false
```
Правильный способ выравнивания здесь такой:

```
if (rookX in ... &&
    rookY in ...) {
    dangerBishop = false
}
```

mglukhikh replied:
Несмотря на то, что в if выполняется один оператор, рекомендуется всё равно ставить фигурные скобки

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -101,7 +114,43 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    // => король под угрозой слона
+    if (abs(bishopX - kingX) == abs(bishopY - kingY))
+    {
+        dangerbishop = true
```
Название переменной состоит из двух слов: danger + bishop. Такие имена по Code Style следует записывать так называемой camelCase нотацией: dangerBishop.

mglukhikh replied:
То же с dangerRook

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -101,7 +114,43 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+        dangerrook = true
+        //если между ладьей и королем стоит слон
+        if (bishopY == kingY && bishopX in min(kingX,rookX)..max(kingX,rookX))
+                    dangerrook = false
```
Здесь тоже немножко неаккуратно выравнивание

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -101,7 +114,43 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    }
+    /////////////////
+    return when {
+        dangerbishop == true && dangerrook == true -> 3 //угроза и от слона, и от ладьи
```
Такие условия записываются просто как `dangerBishop && dangerRook`, запись `== true` здесь избыточна

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +79,27 @@ fun timeForHalfWay(t1: Double, v1: Double,
+    if (rookX2 == kingX || rookY2 == kingY) danger2 = true //есть угроза от второй ладьи
+    if (rookX1 == rookX2) { //если две ладьи и король стоят в один ряд(по вертикали поля)
+        //если между королем и второй ладьей стоит первая ладья, то нет угрозы от второй
+        if (rookY1 in min(rookY2, kingY)..max(rookY2, kingY)) danger2= false
```
Бот сейчас неверно реагирует на подобные ситуации. Но, насколько я вижу глазами, код верен.

mglukhikh replied:
В качестве упражнения: попробуйте добавить тест на подобный случай в **тестовую** функцию whichRookThreatens.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -101,7 +114,43 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    if (abs(bishopX - kingX) == abs(bishopY - kingY))
+    {
+        dangerbishop = true
+        //если между слоном и королем стоит ладья
```
Здесь та же ситуация, что и в задаче с двумя ладьями. Насколько я вижу, код верен

mglukhikh commented:
recheck all

ElenaLaufeyson committed:
Lesson2_task1
Сделана функция triangleKind.

ElenaLaufeyson committed:
Lesson2_task2
Сделана функция isNumberHappy.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +160,18 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
  * Если такой треугольник не существует, вернуть -1.
  */
-fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+fun triangleKind(a: Double, b: Double, c: Double): Int
```
Здесь и далее -- по правилам форматирования кода, принятым в Котлине и Джаве, открывающая фигурная скобка ставится в конце предыдущей строки:

```
fun triangleKind(...): Int {
    ...
}
```

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,7 +17,13 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+{
+    val sum1 = (number/1000 + (number/100)%10)
+    val sum2 = ((number/10)%10 + number%10)
+    if (sum1 == sum2) return true
```
Лучше просто `return sum1 == sum2`

ElenaLaufeyson committed:
Lesson2_task1
Доделан полностью.

ElenaLaufeyson committed:
Lesson2_task2
Доделала полностью.

ElenaLaufeyson committed:
Lesson3_task1
Решила, но еще не полностью задания 3 урока.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена функция fun digitNumber.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена тривиальная задача.

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +167,21 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var cur = 0.0
+    var sinx = 0.0
+    var n = 1
+        while (true) {
```
Весь цикл while нужно сдвинуть на 4 пробела назад

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +167,21 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+            n += 2
+        }
+    return sinx
+}
```
Подумайте, как поведёт себя это решение, если x = 20.0? А при x = 100.0? А при x = 1000.0? Сколько членов ряда потребуется, чтобы достичь требуемой точности? Как будет вести себя погрешность вычислений (помним, что она всегда присутствует при работе с Double)? Можно ли как-то избежать этих неприятных эффектов? Придумайте решение, и добавьте в тестовую функцию один или несколько тестов для подобных ситуаций (сравнивая результат вашей функции с результатом Math,sin с помощью assertEquals).

mglukhikh replied:
Про то, как пишутся тестовые функции -- см. раздел chapter01 в tutorial.

mglukhikh commented:
Второй урок засчитан. Обратите внимание на замечание к функции sin в третьем уроке.

ElenaLaufeyson committed:
Lesson3_task1
Решила одну сложную задачу.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена проблема с функцией sin и добавлена тестовая функция.

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -241,4 +241,28 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+        cur = next
+    }
+    return when {
+        str[n-1] == '0' -> 0
```
Всё проще -- `str[n - 1] - '0'`. Символы тоже можно вычитать (при этом вычитаются их коды), а цифровые символы идут в любой таблице кодировки по порядку от '0' до '9'.

mglukhikh commented on test/lesson3/task1/Tests.kt:
```diff
@@ -133,6 +134,7 @@ class Tests {
         assertEquals(1.0, sin(Math.PI / 2.0, 1e-5), 1e-5)
         assertEquals(0.0, sin(Math.PI, 1e-5), 1e-5)
         assertEquals(-1.0, sin(3.0 * Math.PI / 2.0, 1e-5), 1e-5)
+        assertEquals(Math.sin(20.0), sin(20.0, 1e-5), 1e-5)
```
А что будет, если точность повысить хотя бы до 1e-10? Double, вообще-то, имеет точность в 13-14 десятичных знаков.

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -166,4 +241,28 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+    var next = 0
+    if (n == 1 || n == 2) return 1
+    while (str.length <= n) {
+        str += cur.toString()
```
Здесь задам ещё такой вопрос (бонусный). Как поведёт себя ваша функция при достаточно больших значениях n? Какие факторы могут мешать ей работать правильно? Можно ли что-то сделать с этими факторами, и если можно -- как? Подумайте на эту тему, ответьте мне здесь в комментарии

ElenaLaufeyson replied:
При достаточно больших значениях n число Фибоначчи может выйти за границы типа Int и придется использовать тип Long.
Также можно столкнуться с проблемой переполнения строки или использование большого объема памяти, поэтому лучше рассматривать не всю последовательность целиком, а каждое отдельное число этой последовательности. При данном решении можно даже не использовать строки.

fun fibSequenceDigit(n: Int): Int {

```
  var call = 0 //общая длина всех "проверенных" цифр
  var ccur =0 //длина текущего числа
  var cur = 0 //текущее число
  var i=1 //номер текущего числа
  var ab=0 //какая цифра(по счету) из текущего числа (при заходе числа за границу n) 
```

//понадобится
      var k=0 //последняя цифра текущего числа
      while(true) {
          cur = fib(i) //приравниваем текущему число число последовательности Фиб. с 
//номером i
          ccur = digitNumber(cur) //считаем длину текущего числа(сколько в нем цифр?)
          call+=ccur //увеличиваем длину обработанной пос-ти на длину текущего числа
          if(call==n) return cur%10 //если длина обраб. пос-ти равна n, то
// возвращаем последнюю цифру текущего числа
          if(call > n) { //если обработанная длина "вышла за" n
              ab=n-(call-ccur) //какая цифра данного числа под номером n?
              while(cur>0) { //поиск нужной цифры текущего числа
                      k=cur%10
                      if(ccur==ab) return k
                      cur/=10
                      ccur--
             }
         }
              i++
    }

mglukhikh replied:
Да, ответ верный

mglukhikh commented:
В целом с третьим уроком всё в порядке. Только вот решение проблемы с синусом вы так и не придумали. Плюс есть ещё бонусный вопрос про fibSequenceDigit.

ElenaLaufeyson committed:
Lesson3_task1
Доделала весь третий урок.

mglukhikh commented:
Елена, я сейчас этот Pull Request закрою, а проблему с sin / cos оставлю вам на подумать. Всё остальное решено верно.
